core/internal/filestream/filestreamrequest.go

```diff
@@ -0,0 +1,258 @@
+		eventsLinesToSend:  len(request.EventsLines),
+
+		// TODO: This is not necessarily lightweight.
+		consoleLineRuns: request.ConsoleLines.ToRuns(),
+	}
+
+	reader.isFullRequest = len(reader.consoleLineRuns) <= 1
+
+	return reader
+}
+
+// FileStreamState is state necessary to turn a [FileStreamRequest]
+// into sequence of JSON requests.
+type FileStreamState struct {
+	// HistoryLineNum is the line number where to append history.
+	HistoryLineNum int
+
+	// EventsLineNum is the line number where to append system metrics.
+	EventsLineNum int
+
+	// SummaryLineNum is the line number where to write the summary.
+	//
+	// The same line is always overwritten. The reason we don't solely
+	// update line 0 is for compatibility with previous versions which
+	// appended summaries. The backend only looks at the last line, so
+	// to properly resume such a run we must update the last line.
+	SummaryLineNum int
+
+	// ConsoleLineOffset is an offset to add to all console updates.
+	//
+	// This is used when resuming a run, in which case we want the new
+	// console logs to be appended to the old ones.
+	ConsoleLineOffset int
+}
+
+// GetJSON returns the first JSON request from the sequence represented
+// by the underlying [FileStreamRequest] and updates the [fileStreamState].
+func (r *FileStreamRequestReader) GetJSON(
+	state *FileStreamState,
+) *FileStreamRequestJSON {
+	json := &FileStreamRequestJSON{
+		Files: map[string]offsetAndContent{},
+	}
+
+	if r.historyLinesToSend > 0 {
+		json.Files[HistoryFileName] = offsetAndContent{
+			Offset:  state.HistoryLineNum,
+			Content: r.request.HistoryLines[:r.historyLinesToSend],
+		}
+		state.HistoryLineNum += r.historyLinesToSend
+	}
+	if r.eventsLinesToSend > 0 {
+		json.Files[EventsFileName] = offsetAndContent{
+			Offset:  state.EventsLineNum,
+			Content: r.request.EventsLines[:r.eventsLinesToSend],
+		}
+		state.EventsLineNum += r.eventsLinesToSend
+	}
+	if r.request.LatestSummary != "" {
+		json.Files[SummaryFileName] = offsetAndContent{
+			Offset:  state.SummaryLineNum,
+			Content: []string{r.request.LatestSummary},
+		}
+	}
+	if len(r.consoleLineRuns) > 0 {
+		run := r.consoleLineRuns[0]
+		json.Files[OutputFileName] = offsetAndContent{
+			Offset:  state.ConsoleLineOffset + run.Start,
+			Content: run.Items,
+		}
+	}
+
+	json.Uploaded = make([]string, 0, len(r.request.UploadedFiles))
+	for file := range r.request.UploadedFiles {
+		json.Uploaded = append(json.Uploaded, file)
+	}
+
+	if r.request.Preempting {
+		boolTrue := true
+		json.Preempting = &boolTrue
+	}
+
+	if r.request.Complete && r.isFullRequest {
+		boolTrue := true
+		exitCode := r.request.ExitCode
+		json.Complete = &boolTrue
+		json.ExitCode = &exitCode
+	}
+
+	return json
+}
+
+// Next returns the request minus the data consumed in [GetJSON].
+//
+// The second return value indicates whether the entire request was consumed.
+func (r *FileStreamRequestReader) Next() (*FileStreamRequest, bool) {
```

Comment:
ah yeah, i wish we could avoid the copying, but i agree this code is more readable

Reply:
I spent a lot of time trying, enough to convince myself copying is inevitable :)